File: IRI/subsystems/CargoMech.py
import wpilib

from wpilib import SmartDashboard
from wpilib import SmartDashboard
from wpilib import RobotBase

from ctre import WPI_TalonSRX as Talon

import math

from sim import simComms
import map
import oi

class CargoMech():

    def initialize(self):
        timeout = 15
        SmartDashboard.putNumber("Wrist Power Set", 0)
        SmartDashboard.putNumber("Gravity Power", 0)

        self.F = 0.25
        SmartDashboard.putNumber("F Gain", self.F)

        #self.angle = 0
        #SmartDashboard.putNumber("angle", self.angle)

        self.range = -1200

        self.povPressed = False

        self.maxVolts = 10
        self.wristUpVolts = 4
        self.wristDownVolts = -4
        SmartDashboard.putNumber("Wrist Up Volts", self.wristUpVolts)
        SmartDashboard.putNumber("Wrist Down Volts", self.wristDownVolts)

        self.xbox = oi.getJoystick(2)
        self.joystick0 = oi.getJoystick(0)

        #below is the talon on the intake
        self.intake = Talon(map.intake)
        self.intake.setNeutralMode(2)
        self.intake.configVoltageCompSaturation(self.maxVolts)

        self.intake.configContinuousCurrentLimit(20,timeout) #20 Amps per motor
        self.intake.configPeakCurrentLimit(30,timeout) #30 Amps during Peak Duration
        self.intake.configPeakCurrentDuration(100,timeout) #Peak Current for max 100 ms
        self.intake.enableCurrentLimit(True)

        #Talon motor object created
        self.wrist = Talon(map.wrist)

        if not wpilib.RobotBase.isSimulation():
            self.wrist.configFactoryDefault()

        self.wrist.setInverted(True)

        self.wrist.configVoltageCompSaturation(self.maxVolts)
        self.wrist.setNeutralMode(2)

        self.wrist.configClearPositionOnLimitF(True)

        self.wrist.configContinuousCurrentLimit(20,timeout) #20 Amps per motor
        self.wrist.configPeakCurrentLimit(30,timeout) #30 Amps during Peak Duration
        self.wrist.configPeakCurrentDuration(100,timeout) #Peak Current for max 100 ms
        self.wrist.enableCurrentLimit(True)

    def setIntake(self, mode):
        #Intake/Outtake/Stop, based on the mode it changes the speed of the motor
        if mode == "intake": self.intake.set(0.9)
        elif mode == "outtake": self.intake.set(-0.9)
        else: self.intake.set(0)

    def setWrist(self, mode):
        [setpoint, error] = [10, 5]
        mult = abs(self.getAngle()-50)/100 + 0.5 #increase constant if the arm is not moving enough close to the setpoint
        if mode == "rocket" and self.getAngle() < setpoint-error: self.moveWrist(self.wristDownVolts/ self.maxVolts*mult, mode)
        elif mode == "rocket" and self.getAngle() > setpoint+error: self.moveWrist(self.wristUpVolts/self.maxVolts*mult, mode)
        elif mode == "up": self.moveWrist(self.wristUpVolts/self.maxVolts, mode)
        elif mode == "down": self.moveWrist(self.wristDownVolts/ self.maxVolts, mode)
        elif mode == "rocket" or mode == "gravity": self.moveWrist(0, mode)
        else: self.moveWrist(-self.getGravity(), "stop")

    def moveWrist(self, power, mode):
        self.out = self.wristBounds(power + self.getGravity(), mode)
        self.wrist.set(self.out)

    def wristBounds(self, power, mode):
        angle = self.getAngle()
        if angle > 80 and (mode == "down" or mode == "gravity"): power = 0
        elif angle < -20 and (mode == "up" or mode == "gravity"): power = 0
        elif angle < 10 and mode == "gravity":
            power = self.F #if operator lets go of buttons while at top, arm goes back to upper limit sqitch
        return power

    def getGravity(self):
        return math.sin(math.radians(self.getAngle()))*self.F

    def getAngle(self):
        pos = 0
        if self.wrist.isRevLimitSwitchClosed():
            pos = self.range
            self.wrist.setQuadraturePosition(pos)
        else:
            pos = self.getPosition()
        angle = abs(pos * 115/self.range)
        return (angle-25)

    '''
    def getAngle(self):
        return self.angle
    '''

    def getPosition(self):
        return self.wrist.getQuadraturePosition()

    def periodic(self):
        if self.xbox.getRawAxis(map.intakeCargo)>0.4: self.setIntake("intake")
        elif self.xbox.getRawAxis(map.outtakeCargo)>0.4: self.setIntake("outtake")
        else: self.setIntake("stop")

        if self.xbox.getPOV() > 0 and self.povPressed: self.povPressed = False
        if self.xbox.getPOV() > 0 and not self.povPressed: self.povPressed = True

        if self.povPressed: self.setWrist("rocket")
        elif self.xbox.getRawButton(map.wristUp): self.setWrist("up")
        elif self.xbox.getRawButton(map.wristDown): self.setWrist("down")
        else: self.setWrist("gravity")

    #disables intake
    def disable(self):
        self.setIntake("stop")
        self.setWrist("stop")

    def getNumber(self, key, defVal):
        val = SmartDashboard.getNumber(key, None)
        if val == None:
            val = defVal
            SmartDashboard.putNumber(key, val)
        return val

    def dashboardInit(self): pass

    def dashboardPeriodic(self):
        SmartDashboard.putNumber("Wrist Position", self.wrist.getQuadraturePosition())
        SmartDashboard.putNumber("Wrist Angle" , self.getAngle())
        SmartDashboard.putNumber("Output", self.out)
        self.F = SmartDashboard.getNumber("F Gain", 0)
        self.wristUpVolts = SmartDashboard.getNumber("Wrist Up Volts", 0)
        self.wristDownVolts = SmartDashboard.getNumber("Wrist Down Volts", 0)
        #SmartDashboard.putBoolean("Limit Switch", self.wrist.isFwdLimitSwitchClosed())
        #SmartDashboard.putBoolean("Limit Switch Reverse", self.wrist.isRevLimitSwitchClosed())
        #SmartDashboard.putBoolean("Wrist PinState Quad A", self.wrist.getPinStateQuadA())
        #SmartDashboard.putBoolean("Wrist PinState Quad B", self.wrist.getPinStateQuadB())

        #self.angle = SmartDashboard.getNumber("angle", 0)

File: Thor-java/src/main/java/frc/robot/subsystems/CargoMech.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package frc.robot.subsystems;

import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.SensorCollection;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;

import edu.wpi.first.wpilibj.command.Subsystem;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.OI;
import frc.robot.RobotMap;
import frc.robot.commands.cargo.CargoHumanControl;

/**
 * Subsystem to manipulate the cargo intake/outake arm and rollers.
 */
public final class CargoMech extends Subsystem {

  /**
   * Different modes of operation for intake wheels.
   */
  public enum IntakeMode {
    /** Spin wheels to pull ball in. */
    intake,
    /** Spin wheels to push ball out. */
    outtake,
    /** Stop wheels from spinning. */
    stop;
  }

  private static final boolean DEBUG = false;

  private double F;
  private int range;
  private double maxVolts;
  private double wristUpVolts;
  private double wristDownVolts;
  private WPI_TalonSRX intake;
  private WPI_TalonSRX wrist;
  private boolean wristAtHome;
  private int wristPosition;
  private boolean wristAtBottom;

  /**
   * Constructs new instance and allocates motor controllers and sensors.
   */
  public CargoMech() {
    super("Cargo");
    int timeout = 15;

    this.F = 0.25;

    // this.angle = 0
    // SmartDashboard.putNumber("angle", this.angle)

    this.range = -1200;

    this.maxVolts = 10.0;
    this.wristUpVolts = 4;
    this.wristDownVolts = -4;

    // below is the talon on the intake
    this.intake = new WPI_TalonSRX(RobotMap.intake);
    this.intake.setName("Cargo", "Intake_" + RobotMap.intake);
    this.intake.setNeutralMode(NeutralMode.Brake);
    this.intake.configVoltageCompSaturation(this.maxVolts);

    this.intake.configContinuousCurrentLimit(20, timeout); // #20 Amps per motor
    this.intake.configPeakCurrentLimit(30, timeout); // 30 Amps during Peak Duration
    this.intake.configPeakCurrentDuration(100, timeout); // Peak Current for max 100 ms
    this.intake.enableCurrentLimit(true);

    // Talon motor object created
    this.wrist = new WPI_TalonSRX(RobotMap.wrist);
    this.wrist.setName("Cargo", "Wrist_" + RobotMap.wrist);
    this.wrist.configFactoryDefault();
    this.wrist.setInverted(true);
    this.wrist.configVoltageCompSaturation(this.maxVolts);
    this.wrist.setNeutralMode(NeutralMode.Brake);
    this.wrist.configClearPositionOnLimitF(true, timeout);
    this.wrist.configContinuousCurrentLimit(20, timeout); // 20 Amps per motor
    this.wrist.configPeakCurrentLimit(30, timeout); // 30 Amps during Peak Duration
    this.wrist.configPeakCurrentDuration(100, timeout); // Peak Current for max 100 ms
    this.wrist.enableCurrentLimit(true);

    // Take initial sensor reading
    readSensors();
  }

  /**
   * This method is called once every iteration by the Scheduler (50 times a
   * second).
   */
  public void periodic() {
    super.periodic();
    // Since reading sensor values takes time, only read them once per iteration.
    readSensors();
  }

  /**
   * Set power/direction of intake wheels based on mode passed in.
   * 
   * @param mode The operation you want (intake, outtake or stop).
   */
  public void setIntake(IntakeMode mode) {
    double power = 0.0;
    if (mode == IntakeMode.intake) {
      power = 0.9;
    } else if (mode == IntakeMode.outtake) {
      power = -0.9;
    }
    intake.set(power);
  }

  /**
   * Set power on wrist motor (unchecked - what you set we apply).
   * 
   * @param power Desired power to apply to motor where negative values move wrist
   *              away from home position towards floor (down) and positive values apply
   *              power to return wrist to home position (up).
   */
  public void setWristPower(double power) {
    wrist.set(power);
  }

  /**
   * Immediately turns off power to both wrist and intake motors.
   */
  public void disable() {
    wrist.set(0);
    intake.set(0);
  }

  /**
   * Helper method used to read the sensors just one time each iteration.
   */
  private void readSensors() {
    SensorCollection wristSensors = wrist.getSensorCollection();
    wristAtHome = wristSensors.isRevLimitSwitchClosed();
    wristAtBottom = wristSensors.isFwdLimitSwitchClosed();
    if (wristAtHome) {
      wristPosition = range;
      wristSensors.setQuadraturePosition(range, 0);
    } else {
      wristPosition = wristSensors.getQuadraturePosition();
    }

    // Now that we've read our sensors, lets update some dashboard values as well
    SmartDashboard.putNumber("Wrist Angle", getAngle());
    SmartDashboard.putBoolean("Limit Switch", wristAtBottom);
    SmartDashboard.putBoolean("Limit Switch Reverse", wristAtHome);

    // Add extra SmartDashboard controls and info if DEBUG constant set to true
    if (DEBUG) {
      SmartDashboard.putNumber("Wrist Position", getPosition());
      SmartDashboard.putNumber("Wrist Output", wrist.get());
      F = OI.getSetNumber("F Gain", F);
      wristUpVolts = OI.getSetNumber("Wrist Up Volts", wristUpVolts);
      wristDownVolts = OI.getSetNumber("Wrist Down Volts", wristDownVolts);
      SmartDashboard.putBoolean("Wrist PinState Quad A", wristSensors.getPinStateQuadA());
      SmartDashboard.putBoolean("Wrist PinState Quad B", wristSensors.getPinStateQuadB());
    }
  }

  /**
   * Nominal output voltage to apply to lift wrist arm (before applying gravity).
   * 
   * @return Output voltage to move wrist up. If DEBUG enabled, you can
   *         see/adjust this value on the dashboard.
   */
  public double getWristUpVolts() {
    return wristUpVolts;
  }

  /**
   * Nominal output power to apply to raise wrist arm (before applying gravity).
   * 
   * @return Power in range of [-1.0, +1.0] required to move wrist up. If DEBUG enabled, you can
   *         see/adjust this value on the dashboard.
   */
  public double getWristUpPower() {
    return wristDownVolts / maxVolts;
  }

  /**
   * Nominal output voltage to apply to lower wrist arm (before applying gravity).
   * 
   * @return Voltage to move wrist down. If DEBUG enabled, you can
   *         see/adjust this value on the dashboard.
   */
  public double getWristDownVolts() {
    return wristDownVolts;
  }

  /**
   * Nominal output power to apply to lower wrist arm (before applying gravity).
   * 
   * @return Power in range of [-1.0, +1.0] required to move wrist down. If DEBUG enabled, you can
   *         see/adjust this value on the dashboard.
   */
  public double getWristDownPower() {
    return wristDownVolts / maxVolts;
  }

  /**
   * Indicates if cargo arm is resting in its home position.\
   * 
   * @return true if arm retracted and resting on upper limit switch.
   */
  public boolean isWristAtHome() {
    return wristAtHome;
  }

  /**
   * Indicates if cargo arm is resting in its home position.\
   * 
   * @return true if arm retracted and resting on upper limit switch.
   */
  public boolean isWristAtBottom() {
    return wristAtBottom;
  }

  /**
   * Returns the raw quaduture position of the cargo arm.
   * 
   * @return Integer count from wrist quadrature encoder.
   */
  public int getPosition() {
    return wristPosition;
  }

  /**
   * Returns the current angle of the wrist in degrees.
   * 
   * @param angDegs
   * @return Angle in degrees - 0 when arm straight up, 90 when horizontal. Angle
   *         will be negative when arm is leaning on back rest.
   */
  public double getAngle() {
    double angle = Math.abs(getPosition() * 115 / range);
    return (angle - 25);
  }

  /**
   * Computes a power adjustment based on wrist angle.
   * 
   * @param angDegs Angle in degrees to compute gravity for (0 when arm straight
   *                up, 90 when horizontal).
   * @return Adjustement to power to compensate for load of arm based on angle.
   */
  public double computeGravity(double angDegs) {
    return Math.sin(Math.toRadians(angDegs)) * F;
  }

  @Override
  public void initDefaultCommand() {
    // Set the default command to be human controlled - NOTE
    // may need to pass reference to "this" to command constructor
    // if this method is invoked in the CargoMech constructor.
    setDefaultCommand(new CargoHumanControl());
  }

}
